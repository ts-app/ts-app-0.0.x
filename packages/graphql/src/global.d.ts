declare module 'clear-module'
declare module 'graphql-voyager/middleware'
declare module 'import-fresh'
