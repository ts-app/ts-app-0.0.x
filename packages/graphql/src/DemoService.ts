import { Resolver } from './Resolver'
import { ServiceInfo } from './ServiceInfo'

export class DemoService {
  private anotherNumber: number = 123

  constructor (public count: number = 0) {
  }

  @Resolver({ paramNames: [ 'name', 'message' ] })
  echo (name: string, message: string) {
    return `Hello, ${name}! Message is "${message}"`
  }

  @Resolver({ type: 'mutation' })
  add ({ howMany = 1 }: { howMany?: number }) {
    this.count += howMany
    return this.count
  }

  @Resolver({ type: 'mutation' })
  async promiseToSubtract ({ howMany = 1 }: { howMany?: number }): Promise<number> {
    return new Promise<number>(resolve => {
      setTimeout(() => {
        this.count -= howMany
        resolve(this.count)
      }, 1)
    })
  }

  justUseAnotherNumber () {
    return this.anotherNumber * 3
  }

  @Resolver()
  throwSomeError () {
    throw new Error('simulated error')
  }
}

export class DemoServiceWithServiceInfo implements ServiceInfo {
  info (): { serviceName: string } {
    return {
      serviceName: 'CustomServiceName'
    }
  }

  @Resolver({ paramNames: [ 'name', 'message' ] })
  echo2 (name: string, message: string) {
    return `Hello, ${name}! Message is "${message}" - from echo2`
  }
}
