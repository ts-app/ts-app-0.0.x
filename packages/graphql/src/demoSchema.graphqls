type Query {
  echo (name: String!, message: String!): String
  throwSomeError: String
}

type Mutation {
  add (howMany: Int): Int
  promiseToSubtract (howMany: Int): Int
}
