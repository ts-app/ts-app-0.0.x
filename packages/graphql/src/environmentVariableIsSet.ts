import { ResolverMiddlewareInput } from './ResolverMiddlewareInput'

/**
 * This function is used to enable/disable access to GraphQL resolver based on specified environment
 * variable to be set to '1'.
 *
 * @param {string} name Environment variable name
 * @return {(input: ResolverMiddlewareInput) => boolean}
 */
export const environmentVariableIsSet = (name: string) => {
  const authorized = process.env[ name ] === '1'

  return (input: ResolverMiddlewareInput): boolean => authorized
// tslint:disable-next-line
}
