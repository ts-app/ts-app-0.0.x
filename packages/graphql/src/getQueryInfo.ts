import { SchemaInfo } from './SchemaInfo'

export const getQueryInfo = function (schemaInfo: SchemaInfo) {
  const {
    schema, resolvers: {
      Query, Mutation
    }
  } = schemaInfo

  const resolverQueries = Object.keys(Query)
  const typeDefQueries = Object.keys(schema.getQueryType().getFields())
  const queriesWithoutResolver = typeDefQueries.filter(q => !resolverQueries.includes(q))

  const resolverMutations = Object.keys(Mutation)
  const mutationType = schema.getMutationType()
  const typeDefMutations = Object.keys(mutationType ? mutationType.getFields() : [])
  const mutationsWithoutResolvers = typeDefMutations.filter(m => !resolverMutations.includes(m))

  return { typeDefQueries, typeDefMutations, queriesWithoutResolver, mutationsWithoutResolvers }
}
