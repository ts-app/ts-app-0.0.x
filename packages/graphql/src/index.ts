export * from './demoSchemaDefinition'
export * from './DemoService'
export * from './environmentVariableIsSet'
export * from './mergeSchemaDefinitions'
export * from './mergeTypeDefs'
export * from './Resolver'
export * from './ResolverError'
export * from './ResolverMiddlewareInput'
export * from './ResolverService'
export * from './SchemaDefinition'
export * from './SecurityMiddleware'
export * from './serveSchema'
export * from './ServiceInfo'
export * from './standardSchemaDefinition'
export * from './userIdMatchParam'
