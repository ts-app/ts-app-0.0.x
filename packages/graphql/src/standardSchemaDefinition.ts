import { GraphQLScalarType, Kind } from 'graphql'
import { SchemaDefinition } from './SchemaDefinition'

export const standardSchemaDefinition = (): SchemaDefinition => {
  // http://dev.apollodata.com/tools/graphql-tools/scalars.html
  const resolvers = {
    Date: new GraphQLScalarType({
      name: 'Date',
      description: 'Date custom scalar type',
      parseValue (value) {
        return new Date(value) // value from the client
      },
      serialize (value) {
        return value.getTime() // value sent to the client
      },
      parseLiteral (ast) {
        if (ast.kind === Kind.INT) {
          return parseInt(ast.value, 10) // ast value is always in string format
        }
        return null
      }
    })
  }

  const typeDefs = 'scalar Date'

  return {
    resolvers,
    typeDefs
  }
}
