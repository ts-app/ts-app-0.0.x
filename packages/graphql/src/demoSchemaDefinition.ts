import { loadFile } from '@ts-app/common'
import { DemoService } from './DemoService'
import { ResolverService } from './ResolverService'
import { standardSchemaDefinition } from './standardSchemaDefinition'
import { SchemaDefinition } from './SchemaDefinition'

export const demoSchemaDefinition = (): SchemaDefinition => {
  const resolver = ResolverService.getInstance()
  resolver.registerService(new DemoService())
  const typeDefs = loadFile(`${__dirname}/demoSchema.graphqls`)
  const resolvers = resolver.makeResolvers()
  const standard = standardSchemaDefinition()

  return {
    typeDefs,
    resolvers,
    dependencies: {
      standard
    }
  }
}
