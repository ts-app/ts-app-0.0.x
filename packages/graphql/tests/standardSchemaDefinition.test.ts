import { standardSchemaDefinition } from '../src/standardSchemaDefinition'

describe('standardSchemaDefinition', () => {
  test('works', () => {
    const {
      resolvers, typeDefs
    } = standardSchemaDefinition()

    expect(resolvers).toMatchSnapshot()
    expect(typeDefs).toMatchSnapshot()
  })
})
