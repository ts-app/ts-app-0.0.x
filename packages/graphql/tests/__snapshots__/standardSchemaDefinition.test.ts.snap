// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`standardSchemaDefinition works 1`] = `
Object {
  "Date": "Date",
}
`;

exports[`standardSchemaDefinition works 2`] = `"scalar Date"`;
